### DISCUSSION TITLE
Where are the crazyflie-specific system matrices A and B and additional information in the code?

### DISCUSSION THREAD

**Discussion by hedaniel7** [TinyMPC/TinyMPC#12] [Q&A]
Thank you for sharing your very good research and code.

I am working on a drone practical at the Technical University of Munich using the Crazyflie 2.1:
https://github.com/AlboAlby00/CrazyflieControllers/tree/TinyMPC/crazyflie_controllers
(and your code is one of my last straws to make MPC work and save me before the deadline..)

I want to implement your code in a ROS2 node in a Webots simulation.

In your code you say:
https://github.com/TinyMPC/TinyMPC/blob/main/examples/quadrotor_hovering.cpp
`
// This script is just to show how to use the library, the data for this example is not tuned for our Crazyflie demo. Check the firmware code for more details.
`

The thing is that I can't find anything of your code in the Crazyflie-firmware
https://github.com/search?q=repo%3Abitcraze%2Fcrazyflie-firmware%20tiny&type=code
I also tried to search for the keyword: "MPC" and looked for github branches and couldn't find it.

If you by chance meant your own repository branch https://github.com/TinyMPC/TinyMPC/tree/crazyflie-firmware, I checked `problem_data/quadrotor_20hz_params.hpp` of both main and this branch and the values are at least very similar if not identical.

Where could I find `quadrotor_20hz_params.hpp` tuned for the crazyflie? 

Thank you very much for your time and effort

Greetings from Germany :)
Daniel

**Comment by xkhainguyen** [comment]
Hi there,

Thank you for reaching out!

For a better understanding of the entire workflow (even with the real hardware), we recommend you check [this example](https://github.com/TinyMPC/tinympc-python/blob/main/examples/interactive_quadrotor_ext.ipynb). This should provide you with a sense of modeling as well as MPC implementation. You can see our actual Crazyflie model parameters (we guess-estimated them though), but the remaining MPC part is new and has not been tested on hardware. 

We haven't released our "official" Crazyflie firmware yet but you can check [this repo](https://github.com/RoboticExplorationLab/tinympc-crazyflie-firmware). You can build and flash this firmware and it should run out of the box. This is not with our latest TinyMPC source code. For your particular questions, [this is](https://github.com/RoboticExplorationLab/tinympc-crazyflie-firmware/blob/khai/tracking/examples/controller_tinympc_eigen/src/params_500hz.h) the data we use for the figure-8 tracking demo.

Note that despite using a sloppy quadrotor model, our MPC works quite robustly.

I hope this is helpful. Let us know if you have any other questions. Good luck!
